proposals/p2173.md

```diff
@@ -396,36 +404,83 @@ var n: i32;
 fn F(T:! B) -> T.(A.T) { return n; }
 ```
 
+A purely syntactic check is used to determine if an `=` is specified directly in
+an expression:
+
+-   An `=` constraint is specified directly in its enclosing `where` expression.
+-   If an `=` constraint is specified directly in an operand of an `&` or
+    `(`...`)`, then it is specified directly in that enclosing expression.
+
+In an `impl as C` or `impl T as C` declaration in an interface or named
+constraint, `C` is not allowed to directly specify any `=` constraints:
+
+```
+// Compile-time identity function.
+fn Identity[T:! type](x:! T) -> T { return x; }
+
+interface E {
+  // ❌ Rewrite constraint specified directly.
+  impl as A where .T = .U and .U = i32;
+  // ❌ Rewrite constraint specified directly.
+  impl as type & (A where .T = .U and .U = i32);
+  // ✅ Not specified directly, but does not result
+  // in any rewrites being performed.
+  impl as Identity(A where .T = .U and .U = i32);
+}
+```
+
+The same rules apply to `is` constraints. Note that `.T == U` constraints are
+also not allowed in this context, because the reference to `.T` is rewritten to
+`.Self.T`, and `.Self` is ambiguous.
+
+```
+// ❌ Rewrite constraint specified directly in `is`.
+fn F[T:! A where .U is (A where .T = i32)]();
+
+// ❌ Reference to `.U` in same-type constraint is ambiguous:
+// does this mean the outer or inner `.Self.U`?
+fn G[T:! A where .U is (A where .T == i32)]();
+
+// ✅ Not specified directly, but does not result
+// in any rewrites being performed. Return type
+// is not rewritten to `i32`.
+fn H[T:! type where T is C]() -> T.(A.U);
```

Comment:
I don't think the name `T` is available until the end of the binding:
```suggestion
fn H[T:! type where .Self is C]() -> T.(A.U);
```

Reply:
Fixed.